### PR TITLE
Fix debian package build with docker container

### LINKED ISSUE #678: Docker build for Debian packages is not working
Hello, 
     I wrote a comment here https://github.com/iovisor/bcc/pull/610#issuecomment-242699019 but probably an issue is better suited to handle the problem; I tried today to run the build for the Debian packages (inside docker) on a clean setup (jessie 8.5); checked out the latest release, and tried to run

`docker build -t bcc:jessie -f Dockerfile.debian .
`
It breaks, because of missing dependencies:

```
dpkg-checkbuilddeps: Unmet build dependencies: clang-format | clang-format-3.7 python-netaddr python-pyroute2
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
```

thanks

### PR DESCRIPTION
**Pull request by bobrik**
* python-pyroute2 does not exist for debian jessie
* pythin-netaddr was added to dependencies and it needs to be installed
* depend on clang-format like on every other clang package
* rename dockerfiles to be more explicit
* update ubuntu dockerfile to mimic debian as much as possible
* fake `sudo` in the build container, because we are already root
* add usage instructions to dockerfiles

Related to #657, #610. Fixes #678.

### CODE REVIEW

**Comment by bobrik** [comment]
cc @finelli

**Comment by finelli** [comment]
Hello
    it seems that the script still breaks:

```
Step 5 : RUN ./scripts/build-deb.sh
 ---> Running in 6d14e9ba11a7
+ set -e
+ PARALLEL=1
++ mktemp -d /tmp/debuild.XXXXXX
+ TMP=/tmp/debuild.seXQzD
+ trap cleanup EXIT
+ . scripts/git-tag.sh
+++ git describe --abbrev=0
++ git_tag_latest=v0.1.8
+++ git rev-list v0.1.8.. --count
++ git_rev_count=428
++ git_rev_count=429
+++ git log --pretty=%s -n 1
++ git_subject='Merge pull request #675 from brendangregg/master'
++ release=429
++ [[ 429 != \1 ]]
+++ git log --pretty=%h -n 1
++ release=429.git.4ebb7cf
++ revision=0.1.8
+ git archive HEAD --prefix=bcc/ --format=tar.gz -o /tmp/debuild.seXQzD/bcc_0.1.8.orig.tar.gz
+ pushd /tmp/debuild.seXQzD
/tmp/debuild.seXQzD ~/bcc
+ tar xf bcc_0.1.8.orig.tar.gz
+ cd bcc
+ [[ test = \t\e\s\t ]]
+ dch -b -v 0.1.8-429.git.4ebb7cf 'Merge pull request #675 from brendangregg/master'
+ [[ test = \n\i\g\h\t\l\y ]]
+ DEB_BUILD_OPTIONS='nocheck parallel=1'
+ debuild -us -uc
 dpkg-buildpackage -rfakeroot -D -us -uc
dpkg-buildpackage: warning: using a gain-root-command while being root
dpkg-buildpackage: source package bcc
dpkg-buildpackage: source version 0.1.8-429.git.4ebb7cf
dpkg-buildpackage: source distribution UNRELEASED
dpkg-buildpackage: source changed by root <root@f416997e8b71>
 dpkg-source --before-build bcc
dpkg-buildpackage: host architecture amd64
dpkg-checkbuilddeps: Unmet build dependencies: clang-format | clang-format-3.7 python-pyroute2
dpkg-buildpackage: warning: build dependencies/conflicts unsatisfied; aborting
dpkg-buildpackage: warning: (Use -d flag to override.)
debuild: fatal error at line 1376:
dpkg-buildpackage -rfakeroot -D -us -uc failed
+ cleanup
+ [[ -d /tmp/debuild.seXQzD ]]
+ rm -rf /tmp/debuild.seXQzD
The command '/bin/sh -c ./scripts/build-deb.sh' returned a non-zero code: 29
```

could it be the fact that the build inside the container looks for a specific tag / bcc release ?

**Comment by bobrik** [comment]
It looks like you're building `master` when you should be trying this PR.

**Comment by finelli** [comment]
yes sorry: now it starts compiling, but it ends badly:
```
...
In file included from /usr/lib/llvm-3.8/include/clang/AST/ASTContext.h:18:0,
                 from /usr/lib/llvm-3.8/include/clang/Frontend/ASTUnit.h:18,
                 from /usr/lib/llvm-3.8/include/clang/Frontend/FrontendAction.h:24,
                 from /usr/lib/llvm-3.8/include/clang/CodeGen/CodeGenAction.h:13,
                 from /tmp/debuild.SOcaJe/bcc/src/cc/frontends/clang/loader.cc:34:
/usr/lib/llvm-3.8/include/clang/AST/ASTTypeTraits.h: In member function 'const void* clang::ast_type_traits::DynTypedNode::getMemoizationData() const':
/usr/lib/llvm-3.8/include/clang/AST/ASTTypeTraits.h:250:65: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                ? *reinterpret_cast<void *const *>(Storage.buffer)
                                                                 ^
In file included from /usr/lib/llvm-3.8/include/clang/Frontend/ASTUnit.h:18:0,
                 from /usr/lib/llvm-3.8/include/clang/Frontend/FrontendAction.h:24,
                 from /usr/lib/llvm-3.8/include/clang/CodeGen/CodeGenAction.h:13,
                 from /tmp/debuild.SOcaJe/bcc/src/cc/frontends/clang/loader.cc:34:
/usr/lib/llvm-3.8/include/clang/AST/ASTContext.h: In member function 'const clang::ast_type_traits::DynTypedNode* clang::ASTContext::DynTypedNodeList::begin() const':
/usr/lib/llvm-3.8/include/clang/AST/ASTContext.h:496:13: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
             ->begin();
             ^
/usr/lib/llvm-3.8/include/clang/AST/ASTContext.h: In member function 'const clang::ast_type_traits::DynTypedNode* clang::ASTContext::DynTypedNodeList::end() const':
/usr/lib/llvm-3.8/include/clang/AST/ASTContext.h:503:13: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
             ->end();
             ^
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.9/README.Bugs> for instructions.
src/cc/frontends/clang/CMakeFiles/clang_frontend.dir/build.make:57: recipe for target 'src/cc/frontends/clang/CMakeFiles/clang_frontend.dir/loader.cc.o' failed
make[3]: *** [src/cc/frontends/clang/CMakeFiles/clang_frontend.dir/loader.cc.o] Error 4
make[3]: Leaving directory '/tmp/debuild.SOcaJe/bcc/obj-x86_64-linux-gnu'
CMakeFiles/Makefile2:497: recipe for target 'src/cc/frontends/clang/CMakeFiles/clang_frontend.dir/all' failed
make[2]: *** [src/cc/frontends/clang/CMakeFiles/clang_frontend.dir/all] Error 2
make[2]: Leaving directory '/tmp/debuild.SOcaJe/bcc/obj-x86_64-linux-gnu'
Makefile:130: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/tmp/debuild.SOcaJe/bcc/obj-x86_64-linux-gnu'
dh_auto_build: make -j1 returned exit code 2
debian/rules:12: recipe for target 'build' failed
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2
debuild: fatal error at line 1376:
dpkg-buildpackage -rfakeroot -D -us -uc failed
+ cleanup
+ [[ -d /tmp/debuild.SOcaJe ]]
+ rm -rf /tmp/debuild.SOcaJe
The command '/bin/sh -c ./scripts/build-deb.sh' returned a non-zero code: 29
```

**Comment by bobrik** [comment]
@finelli you're out of memory, make sure to have 4gb to compile it successfully.

**Comment by finelli** [comment]
ok

where is it stated that requirement ?

**Comment by finelli** [comment]
ok, the debs have been built

but when I install them on another host, the examples are not working:

```
# python /usr/share/bcc/examples/tracing/tcpv4connect.py
In file included from <built-in>:317:
<command line>:5:10: fatal error: './include/linux/kconfig.h' file not found
#include "./include/linux/kconfig.h"
         ^
1 error generated.
Traceback (most recent call last):
  File "/usr/share/bcc/examples/tracing/tcpv4connect.py", line 74, in <module>
    b = BPF(text=bpf_text)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 197, in __init__
    raise Exception("Failed to compile BPF module %s" % src_file)
Exception: Failed to compile BPF module 
```

it seems related to this thread https://github.com/iovisor/bcc/issues/397

the machine where I built the packages in march is the same where I installed the deb files, so all the dependencies and paths should be already present 

could it be the option 
`-DBCC_KERNEL_HAS_SOURCE_DIR=1`
was missing from the docker build ?

**Comment by drzaeus77** [comment]
add to whitelist

**Comment by finelli** [comment]
sorry @drzaeus77 I do not understand what "add to whitelist" means

**Comment by drzaeus77** [comment]
If you take a look at the regression fixed in #660 you can see that debian/ubuntu are in conflict. So, yes, the issue is that `-BCC_KERNEL_HAS_SOURCE_DIR=1` is missing, but adding it back in needs to be done in a debian-specific way. If you can find a way to work that in according to the conversation in #657 that doesn't break both, it would be awesome!

**Comment by drzaeus77** [comment]
@finelli no worries, that was a message to the buildbot...see https://wiki.jenkins-ci.org/display/JENKINS/GitHub+pull+request+builder+plugin for a background explanation if you're curious.

File: debian/control [hunk 1]
```diff
@@ -3,7 +3,7 @@ Maintainer: Brenden Blanco <bblanco@plumgrid.com>
 Section: misc
 Priority: optional
 Standards-Version: 3.9.5
-Build-Depends: debhelper (>= 9), cmake, libllvm3.7 | libllvm3.8, llvm-3.7-dev | llvm-3.8-dev, libclang-3.7-dev | libclang-3.8-dev, libelf-dev, bison, flex, libedit-dev, clang-format | clang-format-3.7, python-netaddr, python-pyroute2, luajit, libluajit-5.1-dev
+Build-Depends: debhelper (>= 9), cmake, libllvm3.7 | libllvm3.8, llvm-3.7-dev | llvm-3.8-dev, libclang-3.7-dev | libclang-3.8-dev, libelf-dev, bison, flex, libedit-dev, clang-format-3.7 | clang-format-3.8, python-netaddr, luajit, libluajit-5.1-dev
```
Comment (drzaeus77) [inline debian/control]: Please don't remove `clang-format` from the list.

Comment (bobrik) [inline debian/control]: @drzaeus77 I did not remove `clang-format`, I replaced it with `clang-format-3.7 | clang-format-3.8` to be consistent with other llvm and clang dependencies.

**Comment by goldshtn** [comment]
@drzaeus77 @bobrik Can we get this rebased and merged, or is there still an objection? It's been sitting around since August :(

**Comment by 4ast** [comment]
have been pending for too long. Please resubmit if it's still relevant.

**Pull request closed without merging**